network/src/transport/mod.rs

```diff
@@ -286,7 +287,22 @@ async fn upgrade_outbound<T: TSocket>(
     let mut socket = ctxt
         .noise
         .upgrade_outbound(socket, remote_pubkey, AntiReplayTimestamps::now)
-        .await?;
+        .await
+        .map_err(|err| {
+            if err.should_security_log() {
+                sample!(
+                    SampleRate::Duration(Duration::from_secs(15)),
+                    error!(
+                        SecurityEvent::NoiseHandshake,
+                        NetworkSchema::new(&ctxt.noise.network_context)
+                            .network_address(&addr)
+                            .connection_origin(&origin),
+                        error = %err,
+                    )
+                );
+            }
```

Comment:
@mimoo Is this sufficient for what you want to see out of security logging?  Should we use metrics for these security events?

Reply:
This should be enough, IMO. 
We should have metrics and even alerts depending on the security log, not sure any of this is implemented yet. But for a Handshake failure, we probably don't need alerts unless we are seeing a lot of them, hence the need for metrics, I'd say.

Reply:
cc @AnomalRoil 

what about:

* ClientExpectingDifferentPubkey
* ServerParseClient

At the end of the day, perhaps it really doesn't matter if someone connects to us and makes a mistake? We're supposed to address all edge-cases, we might just want to sample on any type of error and alert when we see too many errors of the same type?

Reply:
I'm actually wondering now, what is a good criteria for security logging? Some of these error types are, as @mimoo pointed out, MUST_FIX programmer errors, which should hopefully get caught in CI. Other than that, there's no clear "this error case can _only_ be a malicious attack" except for maybe `ServerReplayDetected`. We've seen pretty much all other cases as a result of configuration error, and I don't know if I want to spam security-logging when someone doesn't configure their validator properly.

Actually, we probably want different policies here for validator vs vfn... We can definitely be more aggressive about security-logging for validators (the network address shouldn't be known, so this could be early signal that it leaked?) vs. vfn's we probably don't want to security-log anything unless it's a super clear vuln/issue

Reply:
Basically what you want is a field that is the same as `unreachable` but doesn't panic.

We could create a `Schema` that all it does is add the field `MUST_FIX=true` or whatever which would keep consistency across these types of shouldn't happen situations.

Now, this is one where we definitely don't want to panic, so I do support having it just an error.  But, if you do see this log, what would you do?

Reply:
Yeah I think you're right.
We should probably rename security logs -> byzantine logging
If we do this, we can see that noise errors are probably not security logs, but probably something you still want to monitor/alert on at the association level